Add python mode

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#63496 Add python mode**

This PR adds a (private) enable_python_mode context manager.
(see torch/utils/_python_dispatch.py).
enable_python_mode accepts the type of a __torch_dispatch__ object
as its argument. Whenever an operator gets called inside of the
context manager, it dispatches to the __torch_dispatch__ of
the passed-in type.

Example usage:
```
with enable_python_mode(LoggingTensor):
    z = torch.empty([])
    assert isinstance(z, LoggingTensor)
```

There are quite a few changes that were made to support this.

First, we added TorchDispatchTypeObject, a C++ struct that represents the
type of a `__torch_dispatch__` object (e.g. LoggingTensor).
It holds both the PyObject* representing the class and a PyInterpreter*
so we know which Python interpreter it came from.

Next, we updated the concrete_dispatch_fn in python_variable.cpp to accept
a `const std::shared_ptr<TorchDispatchTypeObject>&` argument. When this
is null, dispatching happens as usual. When it is non-null, we prepend
the TorchDispatchTypeObject's PyObject* to the overloaded args list so that
it is considered first for dispatch.

To get that to work, we changed how `handle_torch_dispatch_no_python_arg_parser`
works. The "overloaded args list" previously only consisted of Tensor PyObjects,
but now it can have types in addition to Tensors!
- We renamed `append_overloaded_arg` to `append_overloaded_arg`
- We added a new `append_overloaded_type` that appends a type to
overloaded_args
- We added special handling in `handle_torch_dispatch_no_python_arg_parser`
and `append_overloaded_arg` to handle types in addition to Tensors.

Then, there is PythonMode and PythonModeTLS.
- We reuse the DispatchKey::Python dispatch key as a mode key
- We use PythonMode::enter and PythonMode::exit to enable/disable
DispatchKey::Python and set the PythonModeTLS.
- PythonModeTLS stores a TorchDispatchTypeObject as metadata.
- PythonMode is in libtorch_python, and PythonModeTLS is in ATen.
This split is due to the libtorch_python library boundary (because we need
to save TLS in ATen/ThreadLocalState)
- We modify the PythonFallbackKernel to look up
the relevant TorchDispatchTypeObject (if Python Mode is active) and
dispatch using it.

There are two more miscellaneous changes:
- internal_new_from_data (torch/csrc/utils/tensor_new.cpp) gets an
exclude guard. enable_python_mode currently does not handle
torch.tensor and the exclude guard is to prevent a bug.

Future:
- This PR does not allow for the nesting of Python modes. In the future we
should be able to enable this with a more sane no_dispatch API and by changing
the TLS to a stack. For now I did not need this for CompositeImplicitAutograd testing.

Test Plan:
- new tests

Differential Revision: [D30543236](https://our.internmc.facebook.com/intern/diff/D30543236)

zou3519 committed:
Add python mode

This PR adds a (private) enable_python_mode context manager.
(see torch/utils/_python_dispatch.py).
enable_python_mode accepts the type of a __torch_dispatch__ object
as its argument. Whenever an operator gets called inside of the
context manager, it dispatches to the __torch_dispatch__ of
the passed-in type.

Example usage:
```
with enable_python_mode(LoggingTensor):
    z = torch.empty([])
    assert isinstance(z, LoggingTensor)
```

There are quite a few changes that were made to support this.

First, we added TorchDispatchTypeObject, a C++ struct that represents the
type of a `__torch_dispatch__` object (e.g. LoggingTensor).
It holds both the PyObject* representing the class and a PyInterpreter*
so we know which Python interpreter it came from.
We declare an abstract TorchDispatchTypeObject in TensorImpl.h and an
implementation PythonTorchDispatchTypeObject in python_mode.h due to
the libtorch_python library boundary.

Next, we updated the concrete_dispatch_fn in python_variable.cpp to accept
a `const std::shared_ptr<TorchDispatchTypeObject>&` argument. When this
is null, dispatching happens as usual. When it is non-null, we prepend
the TorchDispatchTypeObject's PyObject* to the overloaded args list so that
it is considered first for dispatch.

To get that to work, we changed how `handle_torch_dispatch_no_python_arg_parser`
works. The "overloaded args list" previously only consisted of Tensor PyObjects,
but now it can have types in addition to Tensors!
- We renamed `append_overloaded_arg` to `append_overloaded_arg`
- We added a new `append_overloaded_type` that appends a type to
overloaded_args
- We added special handling in `handle_torch_dispatch_no_python_arg_parser`
and `append_overloaded_arg` to handle types in addition to Tensors.

Then, there is PythonMode and PythonModeTLS.
- We added a new dispatch key, DispatchKey::PythonMode
- We use PythonMode::enter and PythonMode::exit to enable/disable
DispatchKey::PythonMode and set the PythonModeTLS.
- PythonModeTLS stores a TorchDispatchTypeObject as metadata.
- PythonMode is in libtorch_python, and PythonModeTLS is in ATen.
This split is due to the libtorch_python library boundary (because we need
to save TLS in ATen/ThreadLocalState)
- Finally, we register a fallback to PythonMode that looks up
the relevant TorchDispatchTypeObject and dispatches using it.

There are two more miscellaneous changes:
- enable_python_mode respects torch._C._DisableTorchDispatch.
This PR adds the PythonMode key to things that _DisableTorchDispatch disables.
- internal_new_from_data (torch/csrc/utils/tensor_new.cpp) gets an
exclude guard. enable_python_mode currently does not handle
torch.tensor and the exclude guard is to prevent a bug.

Future:
- This PR does not allow for the nesting of Python modes. In the future we
should be able to enable this with a more sane no_dispatch API and by changing
the TLS to a stack. For now I did not need this for CompositeImplicitAutograd testing.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add python mode"


This PR adds a (private) enable_python_mode context manager.
(see torch/utils/_python_dispatch.py).
enable_python_mode accepts the type of a __torch_dispatch__ object
as its argument. Whenever an operator gets called inside of the
context manager, it dispatches to the __torch_dispatch__ of
the passed-in type.

Example usage:
```
with enable_python_mode(LoggingTensor):
    z = torch.empty([])
    assert isinstance(z, LoggingTensor)
```

There are quite a few changes that were made to support this.

First, we added TorchDispatchTypeObject, a C++ struct that represents the
type of a `__torch_dispatch__` object (e.g. LoggingTensor).
It holds both the PyObject* representing the class and a PyInterpreter*
so we know which Python interpreter it came from.
We declare an abstract TorchDispatchTypeObject in TensorImpl.h and an
implementation PythonTorchDispatchTypeObject in python_mode.h due to
the libtorch_python library boundary.

Next, we updated the concrete_dispatch_fn in python_variable.cpp to accept
a `const std::shared_ptr<TorchDispatchTypeObject>&` argument. When this
is null, dispatching happens as usual. When it is non-null, we prepend
the TorchDispatchTypeObject's PyObject* to the overloaded args list so that
it is considered first for dispatch.

To get that to work, we changed how `handle_torch_dispatch_no_python_arg_parser`
works. The "overloaded args list" previously only consisted of Tensor PyObjects,
but now it can have types in addition to Tensors!
- We renamed `append_overloaded_arg` to `append_overloaded_arg`
- We added a new `append_overloaded_type` that appends a type to
overloaded_args
- We added special handling in `handle_torch_dispatch_no_python_arg_parser`
and `append_overloaded_arg` to handle types in addition to Tensors.

Then, there is PythonMode and PythonModeTLS.
- We added a new dispatch key, DispatchKey::PythonMode
- We use PythonMode::enter and PythonMode::exit to enable/disable
DispatchKey::PythonMode and set the PythonModeTLS.
- PythonModeTLS stores a TorchDispatchTypeObject as metadata.
- PythonMode is in libtorch_python, and PythonModeTLS is in ATen.
This split is due to the libtorch_python library boundary (because we need
to save TLS in ATen/ThreadLocalState)
- Finally, we register a fallback to PythonMode that looks up
the relevant TorchDispatchTypeObject and dispatches using it.

There are two more miscellaneous changes:
- enable_python_mode respects torch._C._DisableTorchDispatch.
This PR adds the PythonMode key to things that _DisableTorchDispatch disables.
- internal_new_from_data (torch/csrc/utils/tensor_new.cpp) gets an
exclude guard. enable_python_mode currently does not handle
torch.tensor and the exclude guard is to prevent a bug.

Future:
- This PR does not allow for the nesting of Python modes. In the future we
should be able to enable this with a more sane no_dispatch API and by changing
the TLS to a stack. For now I did not need this for CompositeImplicitAutograd testing.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add python mode"


This PR adds a (private) enable_python_mode context manager.
(see torch/utils/_python_dispatch.py).
enable_python_mode accepts the type of a __torch_dispatch__ object
as its argument. Whenever an operator gets called inside of the
context manager, it dispatches to the __torch_dispatch__ of
the passed-in type.

Example usage:
```
with enable_python_mode(LoggingTensor):
    z = torch.empty([])
    assert isinstance(z, LoggingTensor)
```

There are quite a few changes that were made to support this.

First, we added TorchDispatchTypeObject, a C++ struct that represents the
type of a `__torch_dispatch__` object (e.g. LoggingTensor).
It holds both the PyObject* representing the class and a PyInterpreter*
so we know which Python interpreter it came from.
We declare an abstract TorchDispatchTypeObject in TensorImpl.h and an
implementation PythonTorchDispatchTypeObject in python_mode.h due to
the libtorch_python library boundary.

Next, we updated the concrete_dispatch_fn in python_variable.cpp to accept
a `const std::shared_ptr<TorchDispatchTypeObject>&` argument. When this
is null, dispatching happens as usual. When it is non-null, we prepend
the TorchDispatchTypeObject's PyObject* to the overloaded args list so that
it is considered first for dispatch.

To get that to work, we changed how `handle_torch_dispatch_no_python_arg_parser`
works. The "overloaded args list" previously only consisted of Tensor PyObjects,
but now it can have types in addition to Tensors!
- We renamed `append_overloaded_arg` to `append_overloaded_arg`
- We added a new `append_overloaded_type` that appends a type to
overloaded_args
- We added special handling in `handle_torch_dispatch_no_python_arg_parser`
and `append_overloaded_arg` to handle types in addition to Tensors.

Then, there is PythonMode and PythonModeTLS.
- We added a new dispatch key, DispatchKey::PythonMode
- We use PythonMode::enter and PythonMode::exit to enable/disable
DispatchKey::PythonMode and set the PythonModeTLS.
- PythonModeTLS stores a TorchDispatchTypeObject as metadata.
- PythonMode is in libtorch_python, and PythonModeTLS is in ATen.
This split is due to the libtorch_python library boundary (because we need
to save TLS in ATen/ThreadLocalState)
- Finally, we register a fallback to PythonMode that looks up
the relevant TorchDispatchTypeObject and dispatches using it.

There are two more miscellaneous changes:
- enable_python_mode respects torch._C._DisableTorchDispatch.
This PR adds the PythonMode key to things that _DisableTorchDispatch disables.
- internal_new_from_data (torch/csrc/utils/tensor_new.cpp) gets an
exclude guard. enable_python_mode currently does not handle
torch.tensor and the exclude guard is to prevent a bug.

Future:
- This PR does not allow for the nesting of Python modes. In the future we
should be able to enable this with a more sane no_dispatch API and by changing
the TLS to a stack. For now I did not need this for CompositeImplicitAutograd testing.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add python mode"


This PR adds a (private) enable_python_mode context manager.
(see torch/utils/_python_dispatch.py).
enable_python_mode accepts the type of a __torch_dispatch__ object
as its argument. Whenever an operator gets called inside of the
context manager, it dispatches to the __torch_dispatch__ of
the passed-in type.

Example usage:
```
with enable_python_mode(LoggingTensor):
    z = torch.empty([])
    assert isinstance(z, LoggingTensor)
```

There are quite a few changes that were made to support this.

First, we added TorchDispatchTypeObject, a C++ struct that represents the
type of a `__torch_dispatch__` object (e.g. LoggingTensor).
It holds both the PyObject* representing the class and a PyInterpreter*
so we know which Python interpreter it came from.
We declare an abstract TorchDispatchTypeObject in TensorImpl.h and an
implementation PythonTorchDispatchTypeObject in python_mode.h due to
the libtorch_python library boundary.

Next, we updated the concrete_dispatch_fn in python_variable.cpp to accept
a `const std::shared_ptr<TorchDispatchTypeObject>&` argument. When this
is null, dispatching happens as usual. When it is non-null, we prepend
the TorchDispatchTypeObject's PyObject* to the overloaded args list so that
it is considered first for dispatch.

To get that to work, we changed how `handle_torch_dispatch_no_python_arg_parser`
works. The "overloaded args list" previously only consisted of Tensor PyObjects,
but now it can have types in addition to Tensors!
- We renamed `append_overloaded_arg` to `append_overloaded_arg`
- We added a new `append_overloaded_type` that appends a type to
overloaded_args
- We added special handling in `handle_torch_dispatch_no_python_arg_parser`
and `append_overloaded_arg` to handle types in addition to Tensors.

Then, there is PythonMode and PythonModeTLS.
- We added a new dispatch key, DispatchKey::PythonMode
- We use PythonMode::enter and PythonMode::exit to enable/disable
DispatchKey::PythonMode and set the PythonModeTLS.
- PythonModeTLS stores a TorchDispatchTypeObject as metadata.
- PythonMode is in libtorch_python, and PythonModeTLS is in ATen.
This split is due to the libtorch_python library boundary (because we need
to save TLS in ATen/ThreadLocalState)
- Finally, we register a fallback to PythonMode that looks up
the relevant TorchDispatchTypeObject and dispatches using it.

There are two more miscellaneous changes:
- enable_python_mode respects torch._C._DisableTorchDispatch.
This PR adds the PythonMode key to things that _DisableTorchDispatch disables.
- internal_new_from_data (torch/csrc/utils/tensor_new.cpp) gets an
exclude guard. enable_python_mode currently does not handle
torch.tensor and the exclude guard is to prevent a bug.

Future:
- This PR does not allow for the nesting of Python modes. In the future we
should be able to enable this with a more sane no_dispatch API and by changing
the TLS to a stack. For now I did not need this for CompositeImplicitAutograd testing.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
This is ready for review now. I'm happy to split this PR into smaller pieces to make it easier to read (please let me know) but the pieces don't make sense by themselves

bdhirsh commented on c10/core/DispatchKey.h:
```diff
@@ -133,6 +133,7 @@ enum class DispatchKey : uint8_t {
   BackendSelect,
 
   Python,
+  PythonMode,
```
I have a more general question - what do you think is the correct interaction between this new feature and the previous usages of `__torch_dispatch__`? Are they meant to be used together? E.g.
```
x = SubclassA(torch.randn([]))
y = SubclassA(torch.randn([]))
        with enable_python_mode(SubclassB):
            z = x + y
            # I think z is a subclassB?
            
```

Part of the reason that I'm wondering is because it looks like `Python` and `PythonMode` share a lot of the same logic. Is there something stopping us from merging them? I guess we would have to make `Python` sometimes function as a modal key, for one. And we'd want to make sure that the `TorchDispatchTypeObject` gets priority when choosing what subclass to dispatch to (which I think it already does)

zou3519 replied:
Yes, they are meant to be used together. The use case of that is if we have something like LoggingTensor (or FiniteTensor), then we may want to log ALL tensor operations, even the ones that happen outside of LoggingTensor.

> Part of the reason that I'm wondering is because it looks like Python and PythonMode share a lot of the same logic. Is there something stopping us from merging them?

Now that you mention it, I think we could just combine the two keys. I didn't have a good reason for separating them; I remember Ed proposed a trick for combining Batched and VmapMode and I didn't want to go through that exercise but we don't need to do that here

zou3519 replied:
>  I guess we would have to make Python sometimes function as a modal key, for one.

Yup

> And we'd want to make sure that the TorchDispatchTypeObject gets priority when choosing what subclass to dispatch to (which I think it already does)

Yup, that already happens

zou3519 replied:
Resolution: Python and PythonMode were collapsed into a single key

ezyang commented:
IMO splitting not necessary

ezyang commented on torch/csrc/autograd/python_mode.cpp:
```diff
@@ -0,0 +1,44 @@
+
+namespace torch { namespace autograd {
+
+PythonTorchDispatchTypeObject::PythonTorchDispatchTypeObject(
```
Looking over the method invocations here, I don't think it's actually necessary to

zou3519 replied:
I'm not sure if I can use PyObject in TensorImpl.h. I could get it to work if we're okay with having it be void* instead

ezyang replied:
we have a forward decl of PyObject for this very purpose :P  c.f. PyInterpreter

```
struct PyInterpreter;
struct C10_API PyInterpreter {
  using name_sig = std::string(const PyInterpreter*);
  using decref_sig = void(const PyInterpreter*, PyObject*);
  using detach_sig =

```

zou3519 replied:
Oh nice, let me go play around with that...

ezyang commented on aten/src/ATen/PythonModeTLS.cpp:
```diff
@@ -0,0 +1,53 @@
+#endif
+
+TORCH_LIBRARY_IMPL(_, PythonMode, m) {
+  m.fallback(torch::CppFunction::makeFromBoxedFunction<&dispatchToPython>());
```
Idly wondering: would it beneficial to collapse Python and PythonMode into a single dispatch key?

zou3519 replied:
Yes, they can be collapsed into a single key. I don't see any reason to keep them separate

zou3519 replied:
Resolution: they were collapsed

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -200,12 +200,28 @@ auto handle_torch_function(PyObject* self, const std::string& func_name, PyObjec
   overloaded_types.reserve(overloaded_args.size());
   for (auto &arg : overloaded_args) {
-    overloaded_types.push_back(py::reinterpret_borrow<py::object>((PyObject *) Py_TYPE(arg.ptr())));
+    overloaded_types.push_back(py::reinterpret_borrow<py::object>(get_type_of_overloaded_arg(arg.ptr())));
```
This seems worse, because you introduced an extra PyType_Check even in places where you know that it's definitely not a type

zou3519 replied:
It is indeed worse

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -328,10 +344,11 @@ auto handle_torch_function_indexing(PyObject* self, PyObject* index, PyObject* v
+  PyObject* obj_type = obj_is_type ? obj : (PyObject*)Py_TYPE(obj);
   for (auto &arg : *overloaded_args) {
-    if (Py_TYPE(obj) == Py_TYPE(arg.ptr())) {
+    if (obj_type == get_type_of_overloaded_arg(arg.ptr())) {
```
Blast, needing to check if it's a type or not really sucks. But now I'm wondering... could we possibly make overloaded_args actually be overloaded_types? (I guess the only thing that would break are legacy cases where people `__torch_function__` non-classmethod)

zou3519 replied:
Is the recommendation for `__torch_function__` now to use it as a classmethod? (and yes, the legacy cases would break). I'm down to change it to make it overloaded_types (in an earlier version of this PR it was actually  `overloaded_types` instead) and keep alive a legacy path for the legacy case

ezyang replied:
yes. But the method spelling does work and I suspect some people would get broken if we switched it.

zou3519 replied:
@ezyang would you be fine if we had two separate paths, one for "overloaded args" and one for "overloaded types"? More concretely, that would look like the following:
- handle_torch_function_no_python_arg_parser uses overloaded_args
- (new!) handle_torch_dispatch_no_python_arg_parser uses overloaded_types
- append_overloaded_arg handles overloaded args
- (new!) append_overloaded_type handles overloaded types

This would involve some code duplication between handle_torch_function_no_python_arg_parser and handle_torch_dispatch_no_python_arg_parser as well as between append_overloaded_arg and append_overloaded_type.

I think this approach is reasonable if we are planning on deprecating the legacy non-classmethod torch_function usages -- when we do that then there won't be code duplication. (is this the plan?)

There's a third alternative that looks something like "we turn overloaded_args from a `vector<py::handle>` to `vector<PyObjOrType>`" where `PyObjOrType` is a struct that holds a `py::handle` and a bool. This solves a lot of issues with performing extra PyType_checks and would be better if we're planning on keeping around the legacy torch_function path around for a while.

ezyang replied:
I guess I'm missing information about how many people are incorrectly doing `__torch_function__` as a non-classmethod. If there aren't many I think we can expire this code very quickly, and then I agree with your analysis. I'm not very keen on doing more complicated things to remove PyType_ checks, would rather just regress performance instead.

zou3519 replied:
I did a quick github search. I found 40 usages of `__torch_function__`, of which [77% are using the legacy style](https://docs.google.com/spreadsheets/d/1M3QIrPPugCmvQWaYH2k8dUKWtcZ5j-_buNHyAToewFU/edit?usp=sharing). So I don't think we can deprecate the non-classmethod `__torch_function__` soon, we probably want to do a full deprecation cycle and help migrate some of the more popular users.

Given that, I'm planning on sticking with the strategy that is in this PR that regresses performance

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -0,0 +1,33 @@
+#
+# NB: Calling an operator inside __torch_dispatch__ does go through
+# __torch_dispatch__ again. Please use _DisableTorchDispatch inside
+# __torch_dispatch__ to prevent infinite recursion.
```
To be fixed!

ezyang reviewed (approved):
good show

albanD commented on aten/src/ATen/PythonModeTLS.cpp:
```diff
@@ -0,0 +1,53 @@
+
+namespace at { namespace impl {
+
+#if !defined(CAFFE2_IS_XPLAT_BUILD) && !defined(C10_MOBILE)
```
Why is this protected by mobile macros?

zou3519 replied:
The OSS mobile build fails unless these are here. I didn't dig into the failures, do you think it is worth it?

zou3519 replied:
I'm experimenting with the builds now

albanD replied:
Well, I'm trying to figure out some other mobile related failures that I can't understand.
This feature is expected not to work on mobile? Can we make it an explicit error?

zou3519 replied:
The feature definitely doesn't work on mobile because (correct me if I'm wrong) there is no python interpreter associated with mobile so there is no way to actually call into the public APIs for python mode

zou3519 replied:
Removed the mobile macros and everything still compiles + tests pass so I'm going to leave it in the current state. If the mobile folks end up wanting to save space by disabling the feature they can go for it in a follow up

albanD commented on aten/src/ATen/ThreadLocalState.cpp:
```diff
@@ -21,6 +21,7 @@ ThreadLocalState::ThreadLocalState(bool keep_grad_mode)
   if (keep_grad_mode_) {
     grad_mode_enabled_ = GradMode::is_enabled();
   }
+  python_mode_state_ = at::impl::PythonModeTLS::get_state();
```
FYI: this will conflict with this stack: https://github.com/pytorch/pytorch/pull/63117 that is landing soon

albanD commented on test/test_python_dispatch.py:
```diff
@@ -50,7 +51,10 @@ def unwrap(e):
             return LoggingTensor(e) if isinstance(e, torch.Tensor) else e
 
-        rs = tree_map(wrap, func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs)))
+        # no_dispatch is only needed if you use enable_python_mode.
```
In the case where the two keys would be merged, this won't be needed anymore right?

zou3519 replied:
We still need no_dispatch to prevent infinite recursion. The reason why we need it is:
- let's say the user called torch.empty
- this dispatches to this torch_dispatch item
- and then inside torch_dispatch we call torch.empty
- torch.empty is going to redispatch to the torch_dispatch, unless we add the no_dispatch

albanD commented on torch/csrc/autograd/python_mode.cpp:
```diff
@@ -0,0 +1,44 @@
+}
+
+PythonTorchDispatchTypeObject::~PythonTorchDispatchTypeObject() {
+  Py_DECREF(data_);
```
What is guaranteeing that the GIL is held here?

zou3519 replied:
The only way to delete one of these objects is to call `PythonMode::exit`*. The only way to call PythonMode::exit is through Python.

zou3519 replied:
Er, I'm not sure if that logic is correct. I updated the code to call pyinterpreter_->decref which does hold the GIL (and doesn't seem to deadlock)

albanD replied:
Given the current implementation yes. But nothing prevents someone from deleting that object from c++ without holding the gil. 
Maybe we want a debug internal assert here just in case we add this in the future?

zou3519 replied:
I updated the code to hold the GIL (via a call to pyinterpreter_->decref)

zou3519 replied:
Yes, updated the code to hold the GIL (via a call to pyinterpreter_->decref)

albanD commented on torch/utils/_python_dispatch.py:
```diff
@@ -0,0 +1,33 @@
+#   is able to selectively disable __torch_dispatch__ of a particular class.
+# - It doesn't work with the tensor constructors (torch.tensor, torch.Tensor)
+@contextlib.contextmanager
+def enable_python_mode(cls) -> Iterator[None]:
```
I am wondering if we want a more descriptive name for this? In particular mentioning that it changes the default Tensor subclass generated by ops?

zou3519 replied:
This isn't being exposed yet so we don't need a better name for it in this PR. I think we do want a more descriptive name (because "mode" only makes sense to us); do you have ideas?

> it changes the default Tensor subclass generated by ops

I wouldn't say it changes the default Tensor subclass, it calls into the `__torch_dispatch__` of the Tensor subclass and then that does whatever it wants. The `__torch_dispatch__` doesn't have to return an instance of the Tensor subclass if it doesn't want to

zou3519 replied:
"dispatch_override"? "omni_dispatch_override"?

albanD replied:
default dispatch maybe?

> This isn't being exposed yet so we don't need a better name for it in this PR.

While I agree with that, it is started to get confusing even for me to have multiple keys/backend systems all called the same :D

zou3519 replied:
I added this to the list of things to follow up on (in the TODO for `torch/utils/_python_dispatch.py`). I want to get this PR in first because it's difficult to land and then we can figure out the naming in the next PR. How does that sound?

albanD replied:
Sure.

zou3519 committed:
Update on "Add python mode"


This PR adds a (private) enable_python_mode context manager.
(see torch/utils/_python_dispatch.py).
enable_python_mode accepts the type of a __torch_dispatch__ object
as its argument. Whenever an operator gets called inside of the
context manager, it dispatches to the __torch_dispatch__ of
the passed-in type.

Example usage:
```
with enable_python_mode(LoggingTensor):
    z = torch.empty([])
    assert isinstance(z, LoggingTensor)
```

There are quite a few changes that were made to support this.

First, we added TorchDispatchTypeObject, a C++ struct that represents the
type of a `__torch_dispatch__` object (e.g. LoggingTensor).
It holds both the PyObject* representing the class and a PyInterpreter*
so we know which Python interpreter it came from.

Next, we updated the concrete_dispatch_fn in python_variable.cpp to accept
a `const std::shared_ptr<TorchDispatchTypeObject>&` argument. When this
is null, dispatching happens as usual. When it is non-null, we prepend
the TorchDispatchTypeObject's PyObject* to the overloaded args list so that
it is considered first for dispatch.

To get that to work, we changed how `handle_torch_dispatch_no_python_arg_parser`
works. The "overloaded args list" previously only consisted of Tensor PyObjects,
but now it can have types in addition to Tensors!
- We renamed `append_overloaded_arg` to `append_overloaded_arg`
- We added a new `append_overloaded_type` that appends a type to
overloaded_args
- We added special handling in `handle_torch_dispatch_no_python_arg_parser`
and `append_overloaded_arg` to handle types in addition to Tensors.

Then, there is PythonMode and PythonModeTLS.
- We reuse the DispatchKey::Python dispatch key as a mode key
- We use PythonMode::enter and PythonMode::exit to enable/disable
DispatchKey::Python and set the PythonModeTLS.
- PythonModeTLS stores a TorchDispatchTypeObject as metadata.
- PythonMode is in libtorch_python, and PythonModeTLS is in ATen.
This split is due to the libtorch_python library boundary (because we need
to save TLS in ATen/ThreadLocalState)
- We modify the PythonFallbackKernel to look up
the relevant TorchDispatchTypeObject (if Python Mode is active) and
dispatch using it.

There are two more miscellaneous changes:
- internal_new_from_data (torch/csrc/utils/tensor_new.cpp) gets an
exclude guard. enable_python_mode currently does not handle
torch.tensor and the exclude guard is to prevent a bug.

Future:
- This PR does not allow for the nesting of Python modes. In the future we
should be able to enable this with a more sane no_dispatch API and by changing
the TLS to a stack. For now I did not need this for CompositeImplicitAutograd testing.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30543236).

zou3519 committed:
Update on "Add python mode"


This PR adds a (private) enable_python_mode context manager.
(see torch/utils/_python_dispatch.py).
enable_python_mode accepts the type of a __torch_dispatch__ object
as its argument. Whenever an operator gets called inside of the
context manager, it dispatches to the __torch_dispatch__ of
the passed-in type.

Example usage:
```
with enable_python_mode(LoggingTensor):
    z = torch.empty([])
    assert isinstance(z, LoggingTensor)
```

There are quite a few changes that were made to support this.

First, we added TorchDispatchTypeObject, a C++ struct that represents the
type of a `__torch_dispatch__` object (e.g. LoggingTensor).
It holds both the PyObject* representing the class and a PyInterpreter*
so we know which Python interpreter it came from.

Next, we updated the concrete_dispatch_fn in python_variable.cpp to accept
a `const std::shared_ptr<TorchDispatchTypeObject>&` argument. When this
is null, dispatching happens as usual. When it is non-null, we prepend
the TorchDispatchTypeObject's PyObject* to the overloaded args list so that
it is considered first for dispatch.

To get that to work, we changed how `handle_torch_dispatch_no_python_arg_parser`
works. The "overloaded args list" previously only consisted of Tensor PyObjects,
but now it can have types in addition to Tensors!
- We renamed `append_overloaded_arg` to `append_overloaded_arg`
- We added a new `append_overloaded_type` that appends a type to
overloaded_args
- We added special handling in `handle_torch_dispatch_no_python_arg_parser`
and `append_overloaded_arg` to handle types in addition to Tensors.

Then, there is PythonMode and PythonModeTLS.
- We reuse the DispatchKey::Python dispatch key as a mode key
- We use PythonMode::enter and PythonMode::exit to enable/disable
DispatchKey::Python and set the PythonModeTLS.
- PythonModeTLS stores a TorchDispatchTypeObject as metadata.
- PythonMode is in libtorch_python, and PythonModeTLS is in ATen.
This split is due to the libtorch_python library boundary (because we need
to save TLS in ATen/ThreadLocalState)
- We modify the PythonFallbackKernel to look up
the relevant TorchDispatchTypeObject (if Python Mode is active) and
dispatch using it.

There are two more miscellaneous changes:
- internal_new_from_data (torch/csrc/utils/tensor_new.cpp) gets an
exclude guard. enable_python_mode currently does not handle
torch.tensor and the exclude guard is to prevent a bug.

Future:
- This PR does not allow for the nesting of Python modes. In the future we
should be able to enable this with a more sane no_dispatch API and by changing
the TLS to a stack. For now I did not need this for CompositeImplicitAutograd testing.

Test Plan:
- new tests

Differential Revision: [D30543236](https://our.internmc.facebook.com/intern/diff/D30543236)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30543236).

zou3519 committed:
Update on "Add python mode"


This PR adds a (private) enable_python_mode context manager.
(see torch/utils/_python_dispatch.py).
enable_python_mode accepts the type of a __torch_dispatch__ object
as its argument. Whenever an operator gets called inside of the
context manager, it dispatches to the __torch_dispatch__ of
the passed-in type.

Example usage:
```
with enable_python_mode(LoggingTensor):
    z = torch.empty([])
    assert isinstance(z, LoggingTensor)
```

There are quite a few changes that were made to support this.

First, we added TorchDispatchTypeObject, a C++ struct that represents the
type of a `__torch_dispatch__` object (e.g. LoggingTensor).
It holds both the PyObject* representing the class and a PyInterpreter*
so we know which Python interpreter it came from.

Next, we updated the concrete_dispatch_fn in python_variable.cpp to accept
a `const std::shared_ptr<TorchDispatchTypeObject>&` argument. When this
is null, dispatching happens as usual. When it is non-null, we prepend
the TorchDispatchTypeObject's PyObject* to the overloaded args list so that
it is considered first for dispatch.

To get that to work, we changed how `handle_torch_dispatch_no_python_arg_parser`
works. The "overloaded args list" previously only consisted of Tensor PyObjects,
but now it can have types in addition to Tensors!
- We renamed `append_overloaded_arg` to `append_overloaded_arg`
- We added a new `append_overloaded_type` that appends a type to
overloaded_args
- We added special handling in `handle_torch_dispatch_no_python_arg_parser`
and `append_overloaded_arg` to handle types in addition to Tensors.

Then, there is PythonMode and PythonModeTLS.
- We reuse the DispatchKey::Python dispatch key as a mode key
- We use PythonMode::enter and PythonMode::exit to enable/disable
DispatchKey::Python and set the PythonModeTLS.
- PythonModeTLS stores a TorchDispatchTypeObject as metadata.
- PythonMode is in libtorch_python, and PythonModeTLS is in ATen.
This split is due to the libtorch_python library boundary (because we need
to save TLS in ATen/ThreadLocalState)
- We modify the PythonFallbackKernel to look up
the relevant TorchDispatchTypeObject (if Python Mode is active) and
dispatch using it.

There are two more miscellaneous changes:
- internal_new_from_data (torch/csrc/utils/tensor_new.cpp) gets an
exclude guard. enable_python_mode currently does not handle
torch.tensor and the exclude guard is to prevent a bug.

Future:
- This PR does not allow for the nesting of Python modes. In the future we
should be able to enable this with a more sane no_dispatch API and by changing
the TLS to a stack. For now I did not need this for CompositeImplicitAutograd testing.

Test Plan:
- new tests

Differential Revision: [D30543236](https://our.internmc.facebook.com/intern/diff/D30543236)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30543236).

zou3519 committed:
Update on "Add python mode"


This PR adds a (private) enable_python_mode context manager.
(see torch/utils/_python_dispatch.py).
enable_python_mode accepts the type of a __torch_dispatch__ object
as its argument. Whenever an operator gets called inside of the
context manager, it dispatches to the __torch_dispatch__ of
the passed-in type.

Example usage:
```
with enable_python_mode(LoggingTensor):
    z = torch.empty([])
    assert isinstance(z, LoggingTensor)
```

There are quite a few changes that were made to support this.

First, we added TorchDispatchTypeObject, a C++ struct that represents the
type of a `__torch_dispatch__` object (e.g. LoggingTensor).
It holds both the PyObject* representing the class and a PyInterpreter*
so we know which Python interpreter it came from.

Next, we updated the concrete_dispatch_fn in python_variable.cpp to accept
a `const std::shared_ptr<TorchDispatchTypeObject>&` argument. When this
is null, dispatching happens as usual. When it is non-null, we prepend
the TorchDispatchTypeObject's PyObject* to the overloaded args list so that
it is considered first for dispatch.

To get that to work, we changed how `handle_torch_dispatch_no_python_arg_parser`
works. The "overloaded args list" previously only consisted of Tensor PyObjects,
but now it can have types in addition to Tensors!
- We renamed `append_overloaded_arg` to `append_overloaded_arg`
- We added a new `append_overloaded_type` that appends a type to
overloaded_args
- We added special handling in `handle_torch_dispatch_no_python_arg_parser`
and `append_overloaded_arg` to handle types in addition to Tensors.

Then, there is PythonMode and PythonModeTLS.
- We reuse the DispatchKey::Python dispatch key as a mode key
- We use PythonMode::enter and PythonMode::exit to enable/disable
DispatchKey::Python and set the PythonModeTLS.
- PythonModeTLS stores a TorchDispatchTypeObject as metadata.
- PythonMode is in libtorch_python, and PythonModeTLS is in ATen.
This split is due to the libtorch_python library boundary (because we need
to save TLS in ATen/ThreadLocalState)
- We modify the PythonFallbackKernel to look up
the relevant TorchDispatchTypeObject (if Python Mode is active) and
dispatch using it.

There are two more miscellaneous changes:
- internal_new_from_data (torch/csrc/utils/tensor_new.cpp) gets an
exclude guard. enable_python_mode currently does not handle
torch.tensor and the exclude guard is to prevent a bug.

Future:
- This PR does not allow for the nesting of Python modes. In the future we
should be able to enable this with a more sane no_dispatch API and by changing
the TLS to a stack. For now I did not need this for CompositeImplicitAutograd testing.

Test Plan:
- new tests

Differential Revision: [D30543236](https://our.internmc.facebook.com/intern/diff/D30543236)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30543236).

albanD reviewed (approved):
LGTM!
